lib/vintage_net_wizard/web/endpoint.ex

```diff
@@ -3,17 +3,58 @@ defmodule VintageNetWizard.Web.Endpoint do
   Supervisor for the Web part of the VintageNet Wizard.
   """
   alias VintageNetWizard.Web.{Router, Socket}
-  alias VintageNetWizard.Backend
-  use Supervisor
+
+  use DynamicSupervisor
 
   @doc false
   def start_link(args) do
-    Supervisor.start_link(__MODULE__, args)
+    DynamicSupervisor.start_link(__MODULE__, args, name: __MODULE__)
+  end
+
+  @doc """
+  Start the web server
+
+  If the web server is started then `{:error, already_started}` is returned.
+
+  Only one server can be running at a time.
+  """
+  @spec start_server() :: GenServer.on_start() | {:error, :already_started}
+  def start_server() do
+    port = Application.get_env(:vintage_net_wizard, :port, 80)
+
+    case DynamicSupervisor.count_children(__MODULE__) do
```

Comment:
If you give the child a name, then you don't need to `count_children` to figure out if it's already running. That would get rid of the race condition between `count_children` and `start_child`.

Reply:
When playing with the code a bit more I did:

```
iex(26)> VintageNetWizard.Web.Endpoint.start_server()

09:55:13.059 [error] Failed to start Ranch listener :hello in :ranch_tcp:listen([cacerts: :..., key: :..., cert: :..., port: 4001]) for reason :eaddrinuse (address already in use)

{:error,
 {:shutdown,
  {:failed_to_start_child, :ranch_acceptors_sup,
   {:listen_error, :hello, :eaddrinuse}}}}

```

Where the error tuple is just returned as a  value, so we don't have to name anything and I don't have to do the checks because if the port is already being listened on it won't start. anyways. It does log the error but does not mess up anything at runtime. Any thoughts on matching on error?

Reply:
I don't think that it is good to rely on a tcp port being in use to force the server to be a singleton. My first reaction when seeing that message would be to look everywhere else to see if I had another server running.

Reply:
How about `:max_children`? https://hexdocs.pm/elixir/1.9.1/DynamicSupervisor.html#init/1-options

Reply:
I like that solution!